File: people/thom/test/bak/ptdmain.c
#include <stdio.h>

do_ptd (char *mdat, char *cdat, double *in1)
{
	FILE	*fout, *gout;
	int 	i, j, k;
	double	out1 [1][3][4], out2 [1], out3 [4], out4 [3][4]; 

    if ((fout = fopen (mdat, "w"))==NULL)
    {
        printf("Cannot open %s\n",mdat);
        exit(0);
    }

	TAKE1_M (in1, out1);
	for (i=0; i<1; i++) {
		for (j=0; j<3; j++) {
			for (k=0; k<4; k++) {
				fprintf(fout, "%f ",out1[i][j][k]);
			}
			fprintf (fout, "\n");
		}
		fprintf(fout, "\n");
	} 
	fprintf (fout, "\n\n");
	DROP_M (in1, out1);
	for (i=0; i<1; i++) {
		for (j=0; j<3; j++) {
			for (k=0; k<4; k++) {
				fprintf(fout, "%f ",out1[i][j][k]);
			}
			fprintf (fout, "\n");
		}
		fprintf(fout, "\n");
	} 
	fprintf (fout, "\n\n");
	TAKE2_M (in1, out2);
	fprintf (fout, "%f ", out2 [0]);
	fprintf (fout, "\n\n");
	DROP2_M (in1, out2);
	fprintf (fout, "%f ", out2 [0]);
	fprintf (fout, "\n\n");
	PTD1_M (in1, out3);
	for (i = 0; i < 4; i++)
		fprintf (fout, "%f ", out3 [i]);
	fprintf (fout, "\n\n");
	PSI1_M (in1, out4);
	for (j=0; j<3; j++) {
		for (k=0; k<4; k++) {
			fprintf(fout, "%f ",out4[j][k]);
		}
		fprintf (fout, "\n");
	} 
	fprintf (fout, "\n\n");
	PSI2_M (in1, out3);
	for (i = 0; i < 4; i++)
		fprintf (fout, "%f ", out3 [i]);
	fprintf (fout, "\n\n");
	fclose (fout);

    if ((fout = fopen (cdat, "w"))==NULL)
    {
        printf("Cannot open %s\n",cdat);
        exit(0);
    }

	TAKE1_C (in1, out1);
	for (i=0; i<1; i++) {
		for (j=0; j<3; j++) {
			for (k=0; k<4; k++) {
				fprintf(fout, "%f ",out1[i][j][k]);
			}
			fprintf (fout, "\n");
		}
		fprintf(fout, "\n");
	} 
	fprintf (fout, "\n\n");
	DROP_C (in1, out1);
	for (i=0; i<1; i++) {
		for (j=0; j<3; j++) {
			for (k=0; k<4; k++) {
				fprintf(fout, "%f ",out1[i][j][k]);
			}
			fprintf (fout, "\n");
		}
		fprintf(fout, "\n");
	} 
	fprintf (fout, "\n\n");
	TAKE2_C (in1, out2);
	fprintf (fout, "%f ", out2 [0]);
	fprintf (fout, "\n\n");
	DROP2_C (in1, out2);
	fprintf (fout, "%f ", out2 [0]);
	fprintf (fout, "\n\n");
	PTD1_C (in1, out3);
	for (i = 0; i < 4; i++)
		fprintf (fout, "%f ", out3 [i]);
	fprintf (fout, "\n\n");
	PSI1_C (in1, out4);
	for (j=0; j<3; j++) {
		for (k=0; k<4; k++) {
			fprintf(fout, "%f ",out4[j][k]);
		}
		fprintf (fout, "\n");
	} 
	fprintf (fout, "\n\n");
	PSI2_C (in1, out3);
	for (i = 0; i < 4; i++)
		fprintf (fout, "%f ", out3 [i]);
	fprintf (fout, "\n\n");
	fclose (fout);
}

main (int argc, char *argv[])
{
	double	in1[24]={5,9,23,-3,5,3,-22,3,5,9,10,-3,9,2,3,5,-2,3,-3,-5,4,2,9,7};

    if (argc != 3)
    {
        printf("usage: %s <moa output file> <c output file>\n",argv[0]);
        exit(0);
    }

	do_ptd (argv[1], argv[2], in1);
}
	

File: people/thom/test/ptdmain.f
PROGRAM ptdmain
    REAL, DIMENSION (0:23) :: in1 = (/ 5,9,23,-3,5,3,-22,3,5,9,10,-3,9,2,3,5,-2,3,-3,-5,4,2,9,7/)
    INTERFACE
        SUBROUTINE do_ptd (mdat, cdat, in1)
            CHARACTER(LEN=*) :: mdat, cdat
            REAL, DIMENSION (0:) :: in1
        END SUBROUTINE do_ptd
    END INTERFACE

    CALL do_ptd ("ptdm.dat", "ptdf.dat", in1)
END PROGRAM ptdmain

SUBROUTINE do_ptd (mdat, cdat, in1)
    CHARACTER(LEN=*) :: mdat, cdat
    REAL, DIMENSION (0:) :: in1
	REAL :: out1 (0:11), out2 (0:0), out3 (0:3), out4 (0:11)
    INTEGER :: j

    INTERFACE 
        SUBROUTINE TAKE1_M(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE TAKE1_M
    END INTERFACE

    INTERFACE 
        SUBROUTINE DROP_M(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE DROP_M
    END INTERFACE

    INTERFACE 
        SUBROUTINE TAKE2_M(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE TAKE2_M
    END INTERFACE

    INTERFACE 
        SUBROUTINE DROP2_M(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE DROP2_M
    END INTERFACE

    INTERFACE 
        SUBROUTINE PTD1_M(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE PTD1_M
    END INTERFACE

    INTERFACE 
        SUBROUTINE PSI1_M(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE PSI1_M
    END INTERFACE

    INTERFACE 
        SUBROUTINE PSI2_M(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE PSI2_M
    END INTERFACE

    INTERFACE 
        SUBROUTINE TAKE1_F90(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE TAKE1_F90
    END INTERFACE

    INTERFACE 
        SUBROUTINE DROP1_F90(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE DROP1_F90
    END INTERFACE

    INTERFACE 
        SUBROUTINE TAKE2_F90(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE TAKE2_F90
    END INTERFACE

    INTERFACE 
        SUBROUTINE DROP2_F90(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE DROP2_F90
    END INTERFACE

    INTERFACE 
        SUBROUTINE PTD1_F90(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE PTD1_F90
    END INTERFACE

    INTERFACE 
        SUBROUTINE PSI1_F90(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE PSI1_F90
    END INTERFACE

    INTERFACE 
        SUBROUTINE PSI2_F90(f90_in1, f90_out)
            REAL, DIMENSION (0:) :: f90_in1
            REAL, DIMENSION (0:) :: f90_out
        END SUBROUTINE PSI2_F90
    END INTERFACE

    OPEN (1, FILE=mdat, STATUS="REPLACE", ACCESS="SEQUENTIAL", ACTION="WRITE")

    PRINT *, "HERE"
    CALL TAKE1_M (in1, out1)
    PRINT *, "HERE"
    DO j=0, 2
        WRITE (1,"(4F8.3)") out1(j*4:j*4+3)
    END DO
    WRITE (1,"(A)") ""

    PRINT *, "HERE"
    CALL DROP_M (in1, out1)
    PRINT *, "HERE"
    DO j=0, 2
        WRITE (1,"(4F8.3)") out1(j*4:j*4+3)
    END DO
    WRITE (1,"(A)") ""

    CALL TAKE2_M (in1, out2)
    WRITE (1,"(F8.3)") out2(0)
    WRITE (1,"(A)") ""

    CALL DROP2_M (in1, out2)
    WRITE (1,"(F8.3)") out2(0)
    WRITE (1,"(A)") ""

    CALL PTD1_M (in1, out3)
    WRITE (1,"(4F8.3)") out3(0:3)
    WRITE (1,"(A)") ""

    PRINT *, "HERE"
    CALL PSI1_M (in1, out4)
    PRINT *, "HERE"
    DO j=0, 2
        WRITE (1,"(4F8.3)") out4(j*4:j*4+3)
    END DO
    WRITE (1,"(A)") ""

    CALL PSI2_M (in1, out3)
    WRITE (1,"(4F8.3)") out3(0:3)
    WRITE (1,"(A)") ""


    OPEN (1, FILE=cdat, STATUS="REPLACE", ACCESS="SEQUENTIAL", ACTION="WRITE")

    PRINT *, "HERE"
    CALL TAKE1_F90 (in1, out1)
    PRINT *, "HERE"
    DO j=0, 2
        WRITE (1,"(4F8.3)") out1(j*4:j*4+3)
    END DO
    WRITE (1,"(A)") ""

    PRINT *, "HERE"
    CALL DROP1_F90 (in1, out1)
    PRINT *, "HERE"
    DO j=0, 2
        WRITE (1,"(4F8.3)") out1(j*4:j*4+3)
    END DO
    WRITE (1,"(A)") ""

    CALL TAKE2_F90 (in1, out2)
    WRITE (1,"(F8.3)") out2(0)
    WRITE (1,"(A)") ""

    CALL DROP2_F90 (in1, out2)
    WRITE (1,"(F8.3)") out2(0)
    WRITE (1,"(A)") ""

    CALL PTD1_F90 (in1, out3)
    WRITE (1,"(4F8.3)") out3(0:3)
    WRITE (1,"(A)") ""

    PRINT *, "HERE"
    CALL PSI1_F90 (in1, out4)
    PRINT *, "HERE"
    DO j=0, 2
        WRITE (1,"(4F8.3)") out4(j*4:j*4+3)
    END DO
    WRITE (1,"(A)") ""

    CALL PSI2_F90 (in1, out3)
    WRITE (1,"(4F8.3)") out3(0:3)
    WRITE (1,"(A)") ""

	CLOSE (1)
END SUBROUTINE do_ptd
